src/irr/video/CDerivativeMapCreator.cpp

```diff
@@ -0,0 +1,155 @@
+        gldriver->extGlDeleteShader(cs);
+        gldriver->extGlDeleteProgram(program);
+        return 0;
+    }
+
+    gldriver->extGlAttachShader(program, cs);
+    gldriver->extGlLinkProgram(program);
+
+    //check linking errors
+    success = 0;
+    gldriver->extGlGetProgramiv(program, GL_LINK_STATUS, &success);
+    if (success == GL_FALSE)
+    {
+        gldriver->extGlGetProgramInfoLog(program, sizeof(infoLog), nullptr, infoLog);
+        os::Printer::log("CS LINK ERROR:\n", infoLog, ELL_ERROR);
+        gldriver->extGlDeleteShader(cs);
+        gldriver->extGlDeleteProgram(program);
+        return 0;
+    }
+
+    return program;
+}
+
+namespace
+{
+    constexpr const char* DERIV_MAP_FROM_BUMP_MAP_CS_SRC = R"(
+#version 450 core
+
+layout (local_size_x = 16, local_size_y = 16) in;
+
+layout (binding = 7) uniform sampler2D bumpMapSampler;
+layout (binding = 0, rg8_snorm) uniform image2D derivativeMapImage;
+
+layout (location = 0) uniform float uHeightScaleFactor;
+
+shared float smem[324];//18*18
+
+int getAddr(in ivec2 threadID)
+{
+	return 18*(threadID.y+1) + threadID.x+1;
+}
+
+void main()
+{
+	const ivec2 G_IDX = ivec2(gl_GlobalInvocationID.xy);
+	
+	const ivec2 bumpMapSz = textureSize(bumpMapSampler, 0);
+
+	const ivec2 LC_IDX = ivec2(gl_LocalInvocationID.xy);
+	smem[getAddr(LC_IDX)] = texelFetch(bumpMapSampler, G_IDX, 0).x * uHeightScaleFactor;
+	if (LC_IDX.x == 0 || LC_IDX.x == 15) // TODO how not to use if-statements??
+	{
+		ivec2 offset = ivec2(mix(ivec2(1, 0), ivec2(-1, 0), LC_IDX.x == 0));
+		smem[getAddr(LC_IDX+offset)] = texelFetch(bumpMapSampler, G_IDX+offset, 0).x * uHeightScaleFactor;
+	}
+	if (LC_IDX.y == 0 || LC_IDX.y == 15)
+	{
+		ivec2 offset = ivec2(mix(ivec2(0, 1), ivec2(0, -1), LC_IDX.y == 0));
+		smem[getAddr(LC_IDX+offset)] = texelFetch(bumpMapSampler, G_IDX+offset, 0).x * uHeightScaleFactor;
+	}
+	
+	barrier();
+	memoryBarrierShared();
+	
+	vec2 d = vec2(
+		smem[getAddr(LC_IDX+ivec2(1, 0))] - smem[getAddr(LC_IDX-ivec2(1, 0))],
+		smem[getAddr(LC_IDX+ivec2(0, 1))] - smem[getAddr(LC_IDX-ivec2(0, 1))]
+	) * 0.5 * vec2(bumpMapSz);
+	
+	if (all(lessThan(G_IDX, bumpMapSz)))
+		imageStore(derivativeMapImage, G_IDX, vec4(d, 0.0, 0.0));
+}
+)";
+}
+
+CDerivativeMapCreator::~CDerivativeMapCreator()
+{
+    static_cast<video::COpenGLDriver*>(m_driver)->extGlDeleteProgram(m_deriv_map_gen_cs);
+}
+
+CDerivativeMapCreator::CDerivativeMapCreator(video::IVideoDriver* _driver) : m_driver(_driver)
+{
+    m_deriv_map_gen_cs = this->createComputeShader(DERIV_MAP_FROM_BUMP_MAP_CS_SRC);
+}
+
+video::IVirtualTexture* CDerivativeMapCreator::createDerivMapFromBumpMap(video::IVirtualTexture* _bumpMap, float _heightFactor, bool _texWrapRepeat) const
+{
+    const uint32_t* derivMap_sz = _bumpMap->getSize();
+    video::ITexture* derivMap = m_driver->createGPUTexture(video::ITexture::ETT_2D, derivMap_sz, 1u+uint32_t(std::floor(std::log2(float(core::max_(derivMap_sz[0], derivMap_sz[1]))))), asset::EF_R8G8_SNORM);
```

Comment:
how does this resolve 
https://github.com/buildaworldnet/IrrlichtBAW/pull/333#discussion_r298715523 
?

Reply:
why do you want `ceil` and not `floor`? + i've just taken the code from `COpenGL2DTexture`

Reply:
ah yeah its 1+floor() so ok